File: 1636/SortArrayByIncreasingFrequency.java
/*
https://leetcode.com/problems/sort-array-by-increasing-frequency/description/?envType=daily-question&envId=2024-07-23
*/

class Solution {
    public int[] frequencySort(int[] nums) {
        HashMap<Integer, Integer>numberToItsCountMap=new HashMap<>();
        TreeMap<Integer, ArrayList<Integer>>numbersWithFollowingCount=new TreeMap<>();
        int lengthOfNums=nums.length;
        for(int i=0; i<lengthOfNums; ++i){
            if(numberToItsCountMap.containsKey(nums[i])){
                numberToItsCountMap.put(nums[i], numberToItsCountMap.get(nums[i])+1);
            } else numberToItsCountMap.put(nums[i], 1);
        }
        for(Map.Entry<Integer, Integer>entry: numberToItsCountMap.entrySet()){
            if(numbersWithFollowingCount.containsKey(entry.getValue())){
                ArrayList<Integer>currentList=numbersWithFollowingCount.get(entry.getValue());
                currentList.add(entry.getKey());
                numbersWithFollowingCount.put(entry.getValue(), currentList);
            } else{
                ArrayList<Integer>currentList=new ArrayList<>();
                currentList.add(entry.getKey());
                numbersWithFollowingCount.put(entry.getValue(), currentList);
            }
        }
        ArrayList<Integer>resultantListWithRequiredSorting=new ArrayList<>();
        for(Map.Entry<Integer, ArrayList<Integer>>entry: numbersWithFollowingCount.entrySet()){
            if(entry.getValue().size()>=2){
                Collections.sort(entry.getValue());
                int sizeOfCurrentList=entry.getValue().size();
                for(int i=sizeOfCurrentList-1; i>=0; --i){
                    for(int j=0; j<entry.getKey(); ++j){
                        resultantListWithRequiredSorting.add(entry.getValue().get(i));
                    }
                }
            } else{
                for(int j=0; j<entry.getKey(); ++j){
                    resultantListWithRequiredSorting.add(entry.getValue().get(entry.getValue().size()-1));
                }
            }
        }
        int[] finalArray=new int[resultantListWithRequiredSorting.size()];
        for(int i=0; i<resultantListWithRequiredSorting.size(); ++i){
            finalArray[i]=resultantListWithRequiredSorting.get(i);
        }
        return finalArray;
    }
}

File: 1/1000s/16s/1636/SortArrayByIncreasingFrequency.cpp
/*
https://leetcode.com/problems/sort-array-by-increasing-frequency/description/?envType=daily-question&envId=2024-07-23
*/

class Solution {
public:
    vector<int> frequencySort(vector<int>& nums) {
        map<int, int>mp1;
        map<int, vector<int>>mp2;
        for(auto &vals: nums){
            mp1[vals]++;
        }
        for(auto &entries: mp1){
            mp2[entries.second].push_back(entries.first);
        }
        vector<int>res;
        for(auto &entries: mp2){
            if(entries.second.size()>=2){
                sort(entries.second.begin(), entries.second.end());
                int sz=entries.second.size();
                for(int i=sz-1; i>=0; --i){
                    for(int j=0; j<entries.first; ++j){
                        res.push_back(entries.second[i]);
                    }
                }
            } else{
                for(int j=0; j<entries.first; ++j){
                    res.push_back(entries.second.back());
                }
            }
        }
        return res;
    }
};